lib/plugin/opsgenie/client.go

```
@@ -0,0 +1,267 @@
+)
+
+var alertBodyTemplate = template.Must(template.New("alert body").Parse(
+	`{{.User}} requested permissions for roles {{range $index, $element := .Roles}}{{if $index}}, {{end}}{{ . }}{{end}} on Teleport at {{.Created.Format .TimeFormat}}.
+{{if .RequestReason}}Reason: {{.RequestReason}}{{end}}
+{{if .RequestLink}}To approve or deny the request, proceed to {{.RequestLink}}{{end}}
+`,
+))
+var reviewNoteTemplate = template.Must(template.New("review note").Parse(
+	`{{.Author}} reviewed the request at {{.Created.Format .TimeFormat}}.
+Resolution: {{.ProposedState}}.
+{{if .Reason}}Reason: {{.Reason}}.{{end}}`,
+))
+var resolutionNoteTemplate = template.Must(template.New("resolution note").Parse(
+	`Access request has been {{.Resolution}}
+{{if .ResolveReason}}Reason: {{.ResolveReason}}{{end}}`,
+))
+
+// OpsgenieClient is a wrapper around resty.Client.
+type OpsgenieClient struct {
+	client *resty.Client
+	// webProxyURL is the address used when building the bodies of the alerts
+	// allowing links to the access requests to be built
+	webProxyURL      *url.URL
+	clusterName      string
+	defaultSchedules []string
+	Priority         String
+}
+
+type OpsgenieClientConfig struct {
+	APIKey           string
+	APIEndpoint      string
+	defaultSchedules []string
+	Priority         String
+}
+
+func NewOpsgenieClient(conf OpsgenieClientConfig, clusterName, webProxyAddr string) (OpsgenieClient, error) {
+	var (
+		webProxyURL *url.URL
+		err         error
+	)
+	if webProxyAddr != "" {
+		if webProxyURL, err = lib.AddrToURL(webProxyAddr); err != nil {
+			return OpsgenieClient{}, trace.Wrap(err)
+		}
+	}
+
+	client := resty.NewWithClient(&http.Client{
+		Timeout: ogHTTPTimeout,
+		Transport: &http.Transport{
+			MaxConnsPerHost:     ogMaxConns,
+			MaxIdleConnsPerHost: ogMaxConns,
+		},
+	})
+	if conf.APIEndpoint != "" {
+		client.SetHostURL(conf.APIEndpoint)
+	} else {
+		client.SetHostURL("") // TODO: Find reasonable default
```

Is there a default endpoint? Or should we return an error if it's not set?